lib/kdl/include/kdl/parallel.h

```diff
@@ -0,0 +1,80 @@
+/*
+ Copyright 2020 Eric Wasylishen
+
+ Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+ documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+ rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+ persons to whom the Software is furnished to do so, subject to the following conditions:
+
+ The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+ Software.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+ WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+#ifndef KDL_PARALLEL_H
+#define KDL_PARALLEL_H
+
+#include <atomic>
+#include <future> // for std::async
+#include <thread>
+#include <utility> // for std::declval
+#include <vector>
+
+namespace kdl {
+    /**
+     * Applies the given lambda to each element of the input (passing elements as const lvalue references),
+     * and returns a vector of the resulting values, in their original order.
+     * 
+     * The lambda is executed in parallel, using the number of threads returned by std::thread::hardware_concurrency().
+     * 
+     * Because the threads are spawned with std::async(std::launch::async, ...) and no thread pool is used,
+     * there is a relatively large overhead and this should only be used on large/slow to process data sets.
+     *
+     * @tparam T the type of the vector elements
+     * @tparam L the type of the lambda to apply
+     * @param input the vector
+     * @param transform the lambda to apply, must be of type `auto(const T&)`
+     * @return a vector containing the transformed values
+     */
+    template<class T, class L>
+    auto vec_parallel_transform(const std::vector<T>& input, L&& transform) {
```

Comment:
The code here looks correct from what I can tell, but I feel a little bit uneasy about it nonetheless. Here are my reasons:

- Parallel code is hard to get right. Even when it looks correct and there are tests, there are so many more possibilities for it to break that it's hard to vouch for it (esp. for me since I don't have that much experience with writing parallel code).
- This function has a narrow use case because it doesn't use a thread pool. It's hard to determine whether some code will benefit from using it or not, and even then it's probably going to depend on a lot of factors at runtime whether this will outperform single threaded execution.
- It's basically a reimplementation of the parallel std::transform that's in the STL.

I think we should consider using either the parallel version of std::transform that's in the STL (but will only work on certain platforms for now), or we should try to find a small simple library of STL-like parallel algorithms that do use a thread pool internally before we start implementing such a library ourselves.

I'm usually all for writing our own algorithms as a learning exercise, but parallel algorithms scare me a bit ;-).

Reply:
As discussed, let's leave the code in and replace it with `std::transform` when it's supported on our target platforms.